### DISCUSSION TITLE
FastAPI async dependencies with structlog.contextvars

### DISCUSSION THREAD

**Discussion by dabljues** [fastapi/fastapi#5999] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import structlog
from fastapi import Depends, FastAPI, Response

LOGGER = structlog.get_logger()


def bind_a():
    structlog.contextvars.bind_contextvars(a="a")


async def bind_b():
    structlog.contextvars.bind_contextvars(b="b")


api = FastAPI(
    dependencies=[Depends(bind_a), Depends(bind_b)],
)


@api.get("/foo")
def foo():
    LOGGER.info("api:get:foo")
    return Response()
```

### Description

I have a rather small FastAPI application that uses [fastapi-azure-auth](https://intility.github.io/fastapi-azure-auth/) to add authentication to various endpoints. It's done by a FastAPI dependency and it's async. This and other FastAPI application that I wrote didn't use `async` at all (I mean there's no `async` in code on my part).

I didn't want to bother anyone with the specifics, so I created a fully reproducible and minimal example of this above. I even had a dummy async dependency imitating the one from `fastapi-azure-auth`, but I found out it has no effect on the code whatsoever (the log value binding from `bind_a` still doesn't work, no matter if this dependency is here or not). It only made me think of making the `bind_x` dependencies `async` (I had them non-async before) So I removed it. So what is happening:

* I have a synchronous endpoint `/foo`
* I have two dependencies: one async, one not - those are global to the API 
  * they do the same thing - bind a value to the logger (different keys) with `structlog.contextvars`
  * I assume most people are at least familiar with `structlog`, but [here](https://www.structlog.org/en/stable/contextvars.html) are the docs on their use of contextvars, if someone would need to take a look

When I make a  `GET http://localhost:8000/foo` request, the output is:

```sh
2023-02-14 14:14:16 [info     ] api:get:foo                    b=b
```

So, only the value set in async dep (`bind_b`) works. However, if I make the `bind_a` dependency async, the value bound in it (`a=`) appears in the log. So you might say, only the values bound in async deps work.

My question is: what should I do here? I mean, I know I can make each dependency `async`. But is it the right way? What's the cause of that? And even if I do that, why wouldn't I have to change the endpoint itself to be `async` as well? (because it works with the `async` dep just fine)

### Operating System

Linux

### FastAPI Version

0.89.1

### Python Version

3.10.4

**Comment by Kludex** [comment, marked as answer]
> But is it the right way?

It is.

You can read more about it here: https://github.com/encode/starlette/pull/1258

**Reply by dabljues** [reply]
Hmm, I see. So, when I'm doing stuff with contextvars, I have to use async. But then, I'm doing this in a dependency - should I also make the endpoints async as well? I mean "it works" without me doing so, but I'm just curious. Maybe I'm not grasping the relation between a sync endpoint and an async dependency

**Reply by Kludex** [reply]
The problem is when you set the value, not when you use it.

**Reply by dabljues** [reply]
Got it. Thank you very much! I was struggling with this so badly, feels good to finally get the answer!
